python/pyiceberg/io/pyarrow.py

```diff
@@ -498,6 +504,49 @@ def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
     return boolean_expression_visit(expr, _ConvertToArrowExpression())
 
 
+@lru_cache
+def _get_file_format(file_format: FileFormat, **kwargs: Dict[str, Any]) -> ds.FileFormat:
+    if file_format == FileFormat.PARQUET:
+        return ds.ParquetFileFormat(**kwargs)
+    else:
+        raise ValueError(f"Unsupported file format: {file_format}")
+
+
+def _construct_fragment(fs: FileSystem, data_file: DataFile, file_format_kwargs: Dict[str, Any] = EMPTY_DICT) -> ds.Fragment:
+    _, path = PyArrowFileIO.parse_location(data_file.file_path)
+    return _get_file_format(data_file.file_format, **file_format_kwargs).make_fragment(path, fs)
+
+
+def _read_deletes(fs: FileSystem, data_file: DataFile) -> Dict[str, pa.ChunkedArray]:
+    delete_fragment = _construct_fragment(
+        fs, data_file, file_format_kwargs={"dictionary_columns": ("file_path",), "pre_buffer": True, "buffer_size": ONE_MEGABYTE}
+    )
+    table = ds.Scanner.from_fragment(fragment=delete_fragment).to_table()
+    table.unify_dictionaries()
+    return {
+        file.as_py(): table.filter(pc.field("file_path") == file).column("pos")
+        for file in table.column("file_path").chunks[0].dictionary
+    }
+
+
+def _create_positional_deletes_indices(positional_deletes: List[pa.ChunkedArray], fn_rows: Callable[[], int]) -> pa.Array:
+    sorted_deleted = merge(*positional_deletes)
+
+    def generator() -> Generator[int, None, None]:
+        deleted_pos = next(sorted_deleted).as_py()  # type: ignore
+        for pos in range(fn_rows()):
+            if deleted_pos == pos:
+                try:
+                    deleted_pos = next(sorted_deleted).as_py()  # type: ignore
+                except StopIteration:
+                    deleted_pos = -1
+            else:
+                yield pos
+
+    # Filter on the positions
+    return pa.array(generator(), type=pa.int64())
```

Comment:
I haven't look closely and don't fully understand what this is doing, but I am wondering: is this doing something that you _should_ ideally be able to do with pyarrow functionality? (since it's doing some transformation from ChunkedArrays to a single Array)

Is this essentially merging the chunked arrays + "inverting" the positions to delete into positions to keep?

Reply:
Iceberg has two types of deletes:
- Positional deletes
- Equality deletes

This is the logic to handle positional deletes. There can be one or more delete files that tell which positions in the data file are considered deleted. Since there can be more files, we need to merge them. We read the delete file using PyArrow, and then we don't want to create a Python array with all the combined delete positions, therefore we have a generator that combines all the delete files into a single PyArrow array.

Reply:
> Is this essentially merging the chunked arrays + "inverting" the positions to delete into positions to keep?

Yes :) The second part is converting the deleted positions `idx = [1, 3, 4]` into `take = [0, 2, 5, 6, ...]`.

Reply:
Would something like this work?

```
In [20]: positional_deletes = [pa.chunked_array([[1, 3], [5, 7]]), pa.chunked_array([[8], [10, 14]])]

In [21]: merged = pa.chunked_array([arr for carr in positional_deletes for arr in carr.chunks])

In [22]: np.setdiff1d(np.arange(20), merged)
Out[22]: array([ 0,  2,  4,  6,  9, 11, 12, 13, 15, 16, 17, 18, 19])
```

(assuming that `fn_rows()` returns 20)

Reply:
Yes, that would be beautiful. I think it makes a lot of sense to do this in PyArrow (especially when there is no GIL involved) because we load the files in parallel.

Reply:
To be honest, I think you could just already use the numpy variant that I mentioned above (it's shorter/simpler than the current code IMO, conversion to numpy is zero-copy for integers, and numpy should already releas ethe GIL as well)

(I think we are certainly open to have this in pyarrow, but given you can easily do it in numpy, it's probably not a high priority issue. But will answer on the issue you opened on the arrow side. Thanks for that!)